src/equations/serre_green_naghdi_1d.jl

```diff
@@ -0,0 +1,960 @@
+            scale_by_mass_matrix!(tmp, D)
+            ldiv!(dv, factorization, tmp)
+        end
+    end
+
+    return nothing
+end
+
+@inline function prim2cons(q, equations::SerreGreenNaghdiEquations1D)
+    h = waterheight(q, equations)
+    v = velocity(q, equations)
+    b = bathymetry(q, equations)
+
+    hv = h * v
+    return SVector(h, hv, b)
+end
+
+@inline function cons2prim(u, equations::SerreGreenNaghdiEquations1D)
+    h, hv, b = u
+
+    eta = h + b
+    v = hv / h
+    D = equations.eta0 - b
+    return SVector(eta, v, D)
+end
+
+@inline function waterheight_total(q, equations::SerreGreenNaghdiEquations1D)
+    return q[1]
+end
+
+@inline function velocity(q, equations::SerreGreenNaghdiEquations1D)
+    return q[2]
+end
+
+@inline function bathymetry(q, equations::SerreGreenNaghdiEquations1D)
+    D = q[3]
+    return equations.eta0 - D
+end
+
+@inline function waterheight(q, equations::SerreGreenNaghdiEquations1D)
+    return waterheight_total(q, equations) - bathymetry(q, equations)
+end
+
+# The entropy/energy takes the whole `q` for every point in space
+"""
+    energy_total_modified(q_global, equations::SerreGreenNaghdiEquations1D, cache)
+
+Return the modified total energy of the primitive variables `q_global` for the
+[`SerreGreenNaghdiEquations1D`](@ref).
+It contains an additional term containing a
+derivative compared to the usual [`energy_total`](@ref) modeling
+non-hydrostatic contributions. The [`energy_total_modified`](@ref)
+is a conserved quantity (for periodic boundary conditions).
+
+For a [`bathymetry_flat`](@ref) the total energy is given by
+```math
+\\frac{1}{2} g h^2 + \\frac{1}{2} h v^2 + \\frac{1}{6} h^3 v_x^2.
+```
+
+`q_global` is a vector of the primitive variables at ALL nodes.
+`cache` needs to hold the SBP operators used by the `solver`.
+"""
+function energy_total_modified(q_global,
+                               equations::SerreGreenNaghdiEquations1D,
+                               cache)
+    # unpack physical parameters and SBP operator `D`
+    g = equations.gravity
+    (; D, v_x) = cache
+
+    # `q_global` is an `ArrayPartition`. It collects the individual arrays for
+    # the water height `h` and the velocity `v`.
+    h, v = q_global.x
+
+    N = length(v)
+    e = zeros(eltype(q_global), N)
+
+    # 1/2 g h^2 + 1/2 h v^2 + 1/6 h^3 w^2
+    # and + 1/8 h (v b_x)^2 for full bathymetry without mild-slope approximation
+    if D isa PeriodicUpwindOperators
+        mul!(v_x, D.minus, v)
+    else
+        mul!(v_x, D, v)
+    end
+
+    if equations.bathymetry isa BathymetryFlat
+        b_x = cache.tmp
+        fill!(b_x, zero(eltype(b_x)))
+    else
+        (; b, b_x) = cache
+        @. b = equations.eta0 - q_global.x[3]
+        if D isa PeriodicUpwindOperators
+            mul!(b_x, D.central, b)
+        else
+            mul!(b_x, D, b)
+        end
+    end
+
+    @. e = 1 / 2 * g * h^2 + 1 / 2 * h * v^2 + 1 / 6 * h * (-h * v_x + 1.5 * v * b_x)^2
```

Comment:
```suggestion
    @. e = 1 / 2 * g * eta^2 + 1 / 2 * h * v^2 + 1 / 6 * h * (-h * v_x + 1.5 * v * b_x)^2
```
Otherwise this is wrong, isn't it? You would also need to compute h before. Can you please check that `eta` and `h` are used consistently everywhere?